### PR TITLE
vim.paste() fixes

### LINKED ISSUE #11420: bracketed paste during :debug triggers :redraw / hangs paste mode
Given `t-debug.vim`:

```vim
function! F()
  echom 1
  echom 2
endfunction

debug call F()
```

Calling `nvim -u t-debug.vim` and then using bracketed paste (after typing "echo", with "40 + 2" in the clipboard) shows:
```
Entering Debug mode.  Type "cont" to continue.
…/Vcs/neovim/t-debug.vim
line 6: call F()
line 6: redraw
>40 + 2
```

Using enter then again, and `q`:
```
>
>q
Error detected while processing …/Vcs/neovim/t-debug.vim:
line    6:
paste: Error executing lua: vim.lua:229: Keyboard interrupt
>echo
```

It looks like a) the bracketed paste should not trigger a redraw in the first place, and b) it then is stuck in the mode then also (so that `q` gets handled/seen by the Lua paste handler as an interrupt).

### LINKED ISSUE #11846: Can't paste into / search while in V-LINE mode
- `nvim --version`: 0.4.3
- `vim -u DEFAULTS` (version: ) behaves differently?
- Operating system/version: macOS Catalina 10.15.2
- Terminal name/version: Terminal.app
- `$TERM`:

### Steps to reproduce using `nvim -u NORC`

Enter visual line mode and then search for text (with `/`) and then paste stuff from the clipboard.

```
echo elephant | pbcopy
nvim -u NORC
ienter some text
to search
around in
elephant<ESC>ggV/<CMD-v>
```

### Actual behaviour

It puts elephant into the buffer like

```
eelephantnter some text
to search...
```

### Expected behaviour

It should put elephant into the command line as your search query `/elephant` and then V-line should move to the elephant in the text (line 4)

### LINKED ISSUE #12184: Visual bracketed baste should replace the selected text instead of pasting at cursor position
- `nvim --version`: 0.4.3
- `vim -u DEFAULTS` (version: ) behaves differently? no
- Operating system/version: macOS Catalina
- Terminal name/version: iTerm2 and alacritty
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`
Assuming clipboard contains the word `apple`

1. Type any word, such as `test`
2. Go to visual mode and select `test` from left to right
3. Paste using paste function of terminal (for example: CMD+V, or Ctrl+V, or Ctrl+Shift+V) that sends a bracketed paste.

### Actual behaviour
Paste results in `testapple`

### Expected behaviour
Just `apple`

Vim8 exhibits the correct expected behavior, even in the command-line version. Macvim, VimR (GUIs), do the same behavior. For anyone looking for a workaround: remap the key bindings for paste in your terminal, make a visual mode map to receive it, and make it do `"0p`.

### LINKED ISSUE #13431: Paste from system clipboard can move a character to end of file
- `nvim --version`: 0.4.4 (also tested on nightly v0.5.0-898-g8fb786e41
- `vim -u DEFAULTS` (version: 8.2) behaves differently? Yes, vim behaves as expected
- Operating system/version: Archlinux
- Terminal name/version: Kitty, xfce4-terminal
- `$TERM`: xterm-kitty, xterm-256color

### Steps to reproduce using `nvim -u NORC`
Copy all text here: https://hastebin.com/raw/vocekupoku (Ctrl+A then Ctrl+C)
Paste into Neovim using Ctrl+Shift+V

Do the same into vim.

### Actual behaviour
Notice that when pasting into Neovim, there is an additional line at the bottom containing just the letter "g". And in Vim there is no extra character.

Doing a diff between the original text and the text pasted into Neovim shows that the g was actually taken from another part of the file.

Output of diff
```
198c198
< 6-7 f: ffbfffgf
---
> 6-7 f: ffbffff
1000a1001
> g
```

### Expected behaviour
The text would paste in its original form, without moving the g.

### Additional notes
Pasting using `"+P` does not have this problem. The problem only occurs when pasting via `Ctrl+Shift+V`

### LINKED ISSUE #17189: paste: Error executing lua: vim.lua:0: attempt to index a nil
### Neovim version (nvim -v)

NVIM v0.7.0-dev Build type: RelWithDebInfo LuaJIT 2.1.0-beta3 Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569/build/config -I/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569/src -I/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569/.deps/usr/include -I/usr/include -I/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569/build/src/nvim/auto -I/build/neovim-o30hrS/neovim-0.7.0~ubuntu1+git202201240048-7e2ce35e3-dd05b3569/build/include Compiled by buildd@lgw01-amd64-031

### Vim (not Nvim) behaves the same?

not

### Operating system/version

Ubuntu 20.04.1 LTS (Focal Fossa) 4.19.128-microsoft-standard on WSL2 in Windows 11

### Terminal name/version

windows terminal / Version: 1.11.3471.0

### $TERM environment variable

xterm-256color

### Installation

system package manager 

### How to reproduce the issue

1. [Open https://github.com/nvim-treesitter/nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter#language-parsers)
2. copy "TSInstall "
3. nvim -u NONE
4. press :
5. do paste: ctrl+shift+v

### Expected behavior

:TSInstall 

### Actual behavior

image
`paste: Error executing lua: vim.lua:0: attempt to index a nil value^@stack traceback:^@^Ivim.lua: in function <vim.lua:0>`

### PR DESCRIPTION
**Pull request by zeertzjq**
Fix #11420
Fix #11846
Fix #12184
Fix #13431
Fix #17189

Supersedes #12465

- [x] Correctly determine cmdline mode
- [x] Correct streamed pasting at both middle and end of line in Normal and Visual mode
- [x] Add tests for the first one
- [x] Deal with some other modes
- [x] Avoid edge cases when chunk is empty
- [x] Deal with trailing new line at the end of a chunk
- [x] Add tests for linewise Visual mode

Replace mode pasting is too complicated. I'll leave it untouched for now.

I've also converted this into a plugin: <https://github.com/zeertzjq/nvim-paste-fix>.

### CODE REVIEW

**Comment by zeertzjq** [comment]
A test seems to fail because I made paste too fast?

**Comment by zeertzjq** [comment]
A test seems to fail because I removed `normal! a`, which made paste too fast?

**Comment by zeertzjq** [comment]
> A test seems to fail because I removed `normal! a`, which made paste too fast?

I added ` | echo` to fix it.

File: test/functional/terminal/tui_spec.lua [hunk 1]
```diff
@@ -379,7 +379,7 @@ describe('TUI', function()
   end)
 
   it('paste: normal-mode (+CRLF #10872)', function()
-    feed_data(':set ruler')
+    feed_data(':set ruler | echo')
```
Comment (zeertzjq) [inline test/functional/terminal/tui_spec.lua:382]: This change is needed because I removed `normal! a` which clears the command line.

**Comment by zeertzjq** [comment]
This feels really hacky because of the inconsistency of `nvim_put()` cursor placement in Normal mode. Things can become much easier if `nvim_put()` with `follow=true` always places cursor on the last character instead of one character past the last character when possible, but changing this will be a breaking change.

**Comment by zeertzjq** [comment]
I feel using `nvim_put()` is really hacky in Normal mode, although it works well in Insert and Terminal mode. Will using `nvim_buf_set_text()` instead cause problems?

**Comment by zeertzjq** [comment]
BTW I've now split one commit into a separate PR #16602 because it is a TUI fix that will affect tests in this PR.

**Comment by zeertzjq** [comment]
Oh, `nvim_buf_set_text` doesn't add the inserted text into the redo buffer, so it needs to be avoided unless there is no alternative.

**Comment by zeertzjq** [comment]
Surprisingly this works in Normal mode for multibyte characters even though I think it won't work.

**Comment by zeertzjq** [comment]
> Surprisingly this works in Normal mode for multibyte characters even though I think it won't work.

Oh, this doesn't work if cursor is on a multibyte character that is at the end of the line when pasting started.

**Comment by zeertzjq** [comment]
As my eol detection is already very hacky and still only works for single-bytes characters, I'm considering using `nvim_win_set_cursor()` instead.

**Comment by plane** [comment]
I tried grabbing the current version of `vim.paste` from this PR to see if paste would work in replace mode.  It works sometimes, but often fails here:

https://github.com/neovim/neovim/blob/11f7e7af0ecc46d1016bc4d5edb5ff8f2b3e31a6/src/nvim/lua/vim.lua#L247 

This fails when `#lines` is zero (`attempt to index a nil value`), which happens fairly often.

This is on Windows Terminal (Preview 1.12.3472.0) with NVIM v0.7.0-dev+848-gb4fbb9dcf.  I'm pasting in a single-line string about 32 characters long.

**Comment by zeertzjq** [comment]
When using `nvim_win_set_cursor()` it is hard to solve the problem in Visual mode. I guess my last option is changing the value of `'virtualedit'`.

**Comment by zeertzjq** [comment]
I'll just make Replace mode streamed pasting and Virtual Replace mode pasting fail in this PR, as they weren't working correctly before anyway, and I may need to first introduce a new API to implement them.

**Comment by zeertzjq** [comment]
The error handling looks ugly because of the need to restore `'virtualedit'`. I'd better first implement a new API.

**Comment by zeertzjq** [comment]
I may have figured out a way without changing `'virtualedit'`. Replace mode is too hard to fix, I'll leave it untouched in this PR.

**Comment by zeertzjq** [comment]
The check for eol and eof in Visual mode is similar to this:
<https://github.com/neovim/neovim/blob/96aa7dd09779d2f0c52cb4fbe5cc785331a7fbd0/src/nvim/normal.c#L7517-L7525>

**Comment by plane** [comment]
I think it's a good idea to emphasize what this patch *doesn't* fix (pasting in replace mode), just to make sure everyone is aware the paste behavior is still buggy and needs more work.

**Comment by zeertzjq** [comment]
Thanks. I previously left a comment about this, but I seemed to delete it later.

**Comment by zeertzjq** [comment]
Oh, I know what you are referring to. Does Replace mode pasting work a bit better now?

**Comment by zeertzjq** [comment]
Vim now uses the same logic as this for bracketed paste in Visual mode: <https://github.com/vim/vim/commit/7a73252063242e88a6966e9d9ae7ea8beb897518>.
